### PR TITLE
Issue 949: Collapse profile - wip

### LINKED ISSUE #701: Can only save enrollment as draft once
I created a new enrollment and was able to save it as a draft once.  The second time I tried to do so (before completing and submitting it), I got a "server error" in the UI and following stack trace:
```
Caused by: javax.persistence.PersistenceException: org.hibernate.HibernateException: identifier of an instance of gov.medicaid.entities.ProviderProfile was altered from 1439 to 0
        at org.hibernate.jpa.spi.AbstractEntityManagerImpl.convert(AbstractEntityManagerImpl.java:1692)
        at org.hibernate.jpa.spi.AbstractEntityManagerImpl.convert(AbstractEntityManagerImpl.java:1602)
        at org.hibernate.jpa.spi.AbstractEntityManagerImpl.convert(AbstractEntityManagerImpl.java:1608)
        at org.hibernate.jpa.internal.EntityManagerImpl$CallbackExceptionMapperImpl.mapManagedFlushFailure(EntityManagerImpl.java:235)
        at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2967)
        at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2339)
        at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:485)
        at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:316)
        at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:47)
        at org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:37)
        at org.jboss.as.txn.service.internal.tsr.JCAOrderedLastSynchronizationList.beforeCompletion(JCAOrderedLastSynchronizationList.java:116)
        at com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:76)
        at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:368)
        at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:91)
        at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:162)
        at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1200)
        ... 139 more
Caused by: org.hibernate.HibernateException: identifier of an instance of gov.medicaid.entities.ProviderProfile was altered from 1439 to 0
        at org.hibernate.event.internal.DefaultFlushEntityEventListener.checkId(DefaultFlushEntityEventListener.java:64)
        at org.hibernate.event.internal.DefaultFlushEntityEventListener.getValues(DefaultFlushEntityEventListener.java:175)
        at org.hibernate.event.internal.DefaultFlushEntityEventListener.onFlushEntity(DefaultFlushEntityEventListener.java:135)
        at org.hibernate.event.internal.AbstractFlushingEventListener.flushEntities(AbstractFlushingEventListener.java:216)
        at org.hibernate.event.internal.AbstractFlushingEventListener.flushEverythingToExecutions(AbstractFlushingEventListener.java:85)
        at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:38)
        at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1282)
        at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:465)
        at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2963)
        ... 150 more
```

I could continue to fill out and submit the enrollment, but everything entered after the first time I saved as draft was lost.

### PR DESCRIPTION
**Pull request by frankduncan**
This updates the persistence code and how the profile and enrollment are linked, no longer blowing away and rebuilding profiles when saving an enrollment.

This is related to #949, but also fixes #701 

WIP because currently renew isn't working

### CODE REVIEW

**Comment by frankduncan** [comment]
This is ready for review, of which will be review phase one, I'm sure.

File: psm-app/db/changelog/db-changelog-1.0.xml [hunk 1]
```diff
@@ -1815,4 +1815,63 @@
...
+  <changeSet author="" id="issue949-entities">
+    <loadUpdateData file="seeds/20180731_entities.csv"
+      primaryKey="entity_id"
+      tableName="entities" />
+  </changeSet>
+
+  <changeSet author="" id="issue949-enrollments">
+    <loadUpdateData file="seeds/20180731_enrollments.csv"
+      primaryKey="enrollment_id"
+      tableName="enrollemnts" />
```
Comment (jasonaowen) [inline psm-app/db/changelog/db-changelog-1.0.xml]: Whoops, this should be `enrollments`; this is causing the integration tests to fail.

**Comment by frankduncan** [comment]
Jenkins, test this please.

File: psm-app/db/changelog/db-changelog-1.0.xml [hunk 2]
```diff
@@ -1982,4 +1982,23 @@
...
+          type="bigint"
+          defaultValue="0">
+          <constraints nullable="false"/>
+        </column>
+      </addColumn>
+    </rollback>
+  </changeSet>
+
+  <changeSet author="" id="issue949-provider-profiles">
+    <loadUpdateData file="seeds/20180731_provider_profiles.csv"
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply psm-app/db/changelog/db-changelog-1.0.xml]: Ah!  Mistake that was a holdover to before the seed sanity liquibase pr got merged.

File: psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/ProviderEnrollmentServiceBean.java [hunk 3]
```diff
@@ -630,19 +630,23 @@ public ProviderProfile getProviderDetails(
...
+        TypedQuery<ProviderProfile> query =
+            getEm().createQuery(
+                "FROM ProviderProfile p WHERE profileId = :profileId",
+                ProviderProfile.class
+            );
         query.setParameter("profileId", profileId);
-        List rs = query.getResultList();
+        List<ProviderProfile> rs = query.getResultList();
         if (rs.isEmpty()) {
             return null;
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/ProviderEnrollmentServiceBean.java]: Yes, and someday I'll finally write that static analyser that yells at you if you ever use "null" in your codebase, because there's no good reason.

File: psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/ProviderEnrollmentServiceBean.java [hunk 4]
```diff
@@ -2435,10 +2412,6 @@ private boolean profileHasPendingSubmission(long profileId) {
     public List<Note> findNotes(long ticketId) {
         List<Note> pendingNotes = new ArrayList<>();
         pendingNotes.addAll(findNotes(0, ticketId));
-        ProviderProfile profile = getProviderDetailsByTicket(ticketId, false);
-        if (profile.getProfileId() > 0) {
-            pendingNotes.addAll(findNotes(profile.getProfileId(), ticketId));
-        }
         return pendingNotes;
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/ProviderEnrollmentServiceBean.java]: This is handled when I remove the profile_id column from the notes in the other PR.

File: psm-app/db/seeds/20180723_provider_profiles.csv [hunk 5]
```diff
@@ -1,15 +0,0 @@
-control_no,profile_id,ticket_id,works_on_reservation,maintains_own_private_practice,employed_or_contracted_by_group,criminal_conviction,civil_penalty,previous_exclusion,risk_level_code,reference_ticket_id
```
Comment (jasonaowen) [inline psm-app/db/seeds/20180723_provider_profiles.csv]: I think this file needs to stay...

File: psm-app/db/seeds/20180731_provider_profiles.csv [hunk 6]
```diff
@@ -0,0 +1,15 @@
+control_no,profile_id,ticket_id,works_on_reservation,maintains_own_private_practice,employed_or_contracted_by_group,criminal_conviction,civil_penalty,previous_exclusion,risk_level_code
```
Comment (jasonaowen) [inline psm-app/db/seeds/20180731_provider_profiles.csv]: ...and this file needs to be deleted!

File: psm-app/services/src/main/java/gov/medicaid/entities/Enrollment.java [hunk 7]
```diff
@@ -53,7 +53,7 @@
     @Id
     @GeneratedValue(strategy = GenerationType.AUTO)
     @Column(name = "enrollment_id")
-    private long ticketId;
+    private long enrollmentId;
```
Comment (jasonaowen) [inline psm-app/services/src/main/java/gov/medicaid/entities/Enrollment.java]: I have wanted this for so long! Thank you!

File: psm-app/db/changelog/db-changelog-1.0.xml [hunk 8]
```diff
@@ -1995,4 +1995,27 @@
...
+        oldColumnName="profile_id"/>
+      <addColumn tableName="provider_profiles">
+        <column name="profile_id"
+          type="bigint"/>
+      </addColumn>
+    </rollback>
+  </changeSet>
+
+  <changeSet author="" id="issue949-entities">
+    <loadUpdateData file="seeds/20180731_entities.csv"
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply psm-app/db/changelog/db-changelog-1.0.xml]: No, for this one, the entities need to be linked to the profile_id, so we actually need to update the data.

**Comment by frankduncan** [comment]
> would you mind tagging that issue in those commits? And expanding their commit messages a bit with why this is useful?

Done!

File: psm-app/db/changelog/db-changelog-1.0.xml (hunk 8 above)
[the remark this answers is not in the document]
Comment (jasonaowen) [reply psm-app/db/changelog/db-changelog-1.0.xml]: Ah! Sorry I missed that. Would you mind mentioning that in the commit message?

**Comment by frankduncan** [comment]
Alright!  I believe everything is handled.

Thanks a ton for the review @jasonaowen

File: psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/ProviderEnrollmentServiceBean.java [hunk 9]
```diff
@@ -2330,6 +2331,10 @@ private long saveTicket(
             // Currently the system manually handles persisting underlying relationships
             // of the provider profile, so to prevent hibernate from complaining that we're
             // changing IDs out from under it, we clone, delete, readd to update them
+            //
+            // We also want to ensure that the ticket currently being saved is the one the
+            // with which provider profile associates
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/ProviderEnrollmentServiceBean.java]: Yes!  This is one of the foundational properties of this change.  From now on, the most recently interacted with enrollment will be the one with which the profile is linked.  And actually, the future of #949 is that you can no longer do the renew action on an enrollment that you've already taken the renew action.

I think that if someone wants to add an "abort renewal," that should be a specific action with specific code around it, to reset it back to the approved enrollment being the active one.

**Comment by frankduncan** [comment]
Because of the timeline, I cherry-picked a commit from the split-enrollment-profile branch into this one, adding the "isActive" flag to UserRequest and Enrollment.  I think it makes the UI less confusing when looking at #967, since providers can only update/renew the latest versions of their Enrollments, but it's easy to back out if we think it's not worth it to add.

File: psm-app/db/changelog/db-changelog-1.0.xml [hunk 10]
```diff
@@ -2075,18 +2075,32 @@
...
-      FROM provider_profiles p
-      WHERE p.enrollment_id = e.enrollment_id;
-
-      UPDATE entities en
-      SET profile_id = p.profile_id
-      FROM provider_profiles p
-      WHERE p.enrollment_id = en.ticket_id;
-    </sql>
-    <rollback/>
+  <changeSet author="" id="issue949-add-fks">
```
Comment (jasonaowen) [inline psm-app/db/changelog/db-changelog-1.0.xml]: Did you mean to replace the `issue949-migrate-enrollment-profile-ids` changeset here?

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/EnrollmentPageFlowController.java [hunk 11]
```diff
@@ -641,34 +641,34 @@ public ModelAndView showPrint(
     /**
      * Shows the print preview for the given ticket.
      *
-     * @param ticketId the ticket to be exported
+     * @param enrollmentId the ticket to be exported
```
Comment (jasonaowen) [inline psm-app/cms-web/src/main/java/gov/medicaid/controllers/EnrollmentPageFlowController.java]: Would you mind updating the JavaDoc to match throughout? Both the `@param` clauses and the description: "Shows the print preview for the given enrollment." (Or just delete it, because it's not very useful.) (Optional)

**Comment by PaulMorris** [comment]
@jasonaowen I have not actually tested this yet.  I was hoping to get to that last week, but since it didn't look like we would get this in for current demos, I focused on master (and there was plenty to do there).  But I am happy to do some testing with this.

**Comment by frankduncan** [comment]
> One of the issues we've uncovered in past testing was that the "My Profiles" page no longer works as expected; this is fixed (by removing the page) in a subsequent PR (right? which one?)

That was actually fixed by enabling the code needed for #967, so, yay?

> We also talked last week about merging this and fixing any issues with it going forward; that's not the way I typically like to work, but I agree it's probably the right thing to do here.

That was just to make sure we had it for demos.  That time has passed.  No worries about merging it when you feel it's ready!

As for testing, I've done a few applications of draft screen one, draft screen two, draft screen three, then submit, approve then renew, deny then edit.  Each time making sure the database only had one profile and it showed up on the appropriate screens.  But outside of that basic smoke testing along expected success paths, I haven't done much.

**Comment by jasonaowen** [comment]
Okay! @frankduncan @PaulMorris thank you, that helps me understand where we are.

> > One of the issues we've uncovered in past testing was that the "My Profiles" page no longer works as expected; this is fixed (by removing the page) in a subsequent PR (right? which one?)
> 
> That was actually fixed by enabling the code needed for #967, so, yay?

@frankduncan do you mean that #967 is the code that fixes it? Or that the issue is now fixed in this PR, and I just haven't re-tested that part yet?

[On Zulip](https://chat.opentechstrategies.com/#narrow/stream/6-Provider-Screening/subject/Checkins/near/51847), @frankduncan asked me for a rough test plan. We should probably record this somewhere in the repo at some point, but for now, here's what I came up with:

- As provider, create an individual enrollment application (test exists)
- As provider, create an org enrollment application
- As admin, approve each
- As provider, renew each (which we can't do because adding the action isn't part of this PR, right?)
- As provider, update each (same)
- As admin, deny an application
- As provider, print preview an application (currently broken on master)
- As provider, print an application (currently broken on master)
- Double-check that reports look sane
- Test search (which is probably also broken? maybe?)

Bonus points if we create & approve an application for every type of provider, though I certainly don't expect you to do so, much less write tests about it. In particular, though, we should probably double-check:

- PCA
- CTCC
- DME
- CCMHR?
- (I don't remember all the edge cases, but istr that the above might have something special about them)

**Comment by frankduncan** [comment]
> Or that the issue is now fixed in this PR, and I just haven't re-tested that part yet?

Code in this PR, specifically commit 8adb7527

**Comment by kfogel** [comment]
@frankduncan [noted in Zulip chat just now](https://chat.opentechstrategies.com/#narrow/stream/6-Provider-Screening/subject/tests/near/53577) that this should have PR #1003 merged first for safety.

File: psm-app/db/changelog/db-changelog-1.0.xml (hunk 10 above)
Comment (frankduncan) [reply psm-app/db/changelog/db-changelog-1.0.xml]: Huh.  Not sure.  Though if our plan is to blow away the DB for anyone going forward, the previous one isn't needed.  I think leaving it out is fine.

File: psm-app/cms-business-process/src/main/java/gov/medicaid/process/enrollment/ScreeningHandler.java [hunk 12]
```diff
@@ -98,6 +98,9 @@ public void executeWorkItem(WorkItem item, WorkItemManager manager) {
 
         // merge rule changes to the model
         try {
+            /*
+             * This may be needed!!!
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply psm-app/cms-business-process/src/main/java/gov/medicaid/process/enrollment/ScreeningHandler.java]: Thanks for the catch!  Added a comment.

File: psm-app/db/changelog/db-changelog-1.0.xml (hunk 10 above)
Comment (PaulMorris) [reply psm-app/db/changelog/db-changelog-1.0.xml]: Good point about DB blow-away plan.  Works for me to leave it out.

File: psm-app/cms-business-process/src/main/java/gov/medicaid/process/enrollment/ScreeningHandler.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (PaulMorris) [reply psm-app/cms-business-process/src/main/java/gov/medicaid/process/enrollment/ScreeningHandler.java]: Thanks, looks good!
